typing_inspect.py

```diff
@@ -231,37 +248,90 @@ def get_origin(tp):
+            get_parameters(Generic) == ()
+            get_parameters(Union) == ()
+            get_parameters(List[int]) == ()
+
+            get_parameters(Generic[T]) == (T,)
+            get_parameters(Tuple[List[T], List[S_co]]) == (T, S_co)
+            get_parameters(Union[S_co, Tuple[T, T]][int, U]) == (U,)
+            get_parameters(Mapping[T, Tuple[S_co, T]]) == (T, S_co)
+        """
+        if NEW_TYPING:
+            if (isinstance(tp, _GenericAlias) or
+                    isinstance(tp, type) and issubclass(tp, Generic) and
+                    tp is not Generic):
+                return tp.__parameters__
+            return ()
+        if (
+            is_generic_type(tp) or is_union_type(tp) or
+            is_callable_type(tp) or is_tuple_type(tp)
+        ):
+            return tp.__parameters__ if tp.__parameters__ is not None else ()
+        else:
+            return ()
+else:
+    def get_parameters(tp):
+        """Return type parameters of a parameterizable type as a tuple
+        in lexicographic order. Parameterizable types are generic types,
+        unions, tuple types and callable types. Examples::
+
+            get_parameters(int) == ()
+            get_parameters(Generic) == ()
+            get_parameters(Union) == ()
+            get_parameters(List[int]) == ()
+
+            get_parameters(Generic[T]) == (T,)
+            get_parameters(Tuple[List[T], List[S_co]]) == (T, S_co)
+            get_parameters(Union[S_co, Tuple[T, T]][int, U]) == (U,)
+            get_parameters(Mapping[T, Tuple[S_co, T]]) == (T, S_co)
+        """
+        # python <= 3.5.2
+        if is_union_type(tp):
+            params = []
+            for arg in (tp.__union_params__ if tp.__union_params__ is not None else ()):
+                params += get_parameters(arg)
+            return tuple(params)
+        elif is_tuple_type(tp):
+            params = []
+            for arg in (tp.__tuple_params__ if tp.__tuple_params__ is not None else ()):
+                params += get_parameters(arg)
+            return tuple(params)
+        elif is_generic_type(tp):
+            params = []
+            base_params = tp.__parameters__
+            if base_params is None:
+                return ()
+            for bp_ in base_params:
+                for bp in (get_args(bp_) if is_tuple_type(bp_) else (bp_, )):
+                    if _has_type_var(bp) and not isinstance(bp, TypeVar):
+                        raise TypeError(
+                            "Cannot inherit from a generic class "
+                            "parameterized with "
+                            "non-type-variable %s" % bp)
+                    if params is None:
+                        params = []
+                    if bp not in params:
+                        params.append(bp)
+            if params is not None:
+                return tuple(params)
+            else:
+                return ()
+        else:
+            return ()
```

Comment:
I think instead of havingg the huge if/else with duplicated docstring you can put just the function body within the if/else.

Reply:
ok. this is less optimal (cost of test at each execution) but I guess this is marginal